ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -121,30 +117,48 @@ public TransactionImple recoverTransaction(Uid actId)
 
 		if (recovered.baseXid() == null)
 		    throw new IllegalArgumentException();
-		
+
 		/*
-		 * Is the transaction already in the list? This may be the case because
+		 * Is the transaction already in the map? This may be the case because
 		 * we scan the object store periodically and may get Uids to recover for
 		 * transactions that are progressing normally, i.e., do not need
-		 * recovery. In which case, we need to ignore them.
+		 * recovery. In which case, we need to ignore them:
+		 *
+		 * ie calling addImportedTransaction with a non null value for recovered will
+		 * call recovered.recordTransaction()
 		 */
 
-		TransactionImple tx = (TransactionImple) _transactions.get(recovered
-				.baseXid());
-
-		if (tx == null)
-		{
+		return addImportedTransaction(recovered, recovered.baseXid(), null, 0);
+	}
 
-			Xid baseXid = recovered.baseXid();
-			_transactions.put(new SubordinateXidImple(baseXid), recovered);
-			recovered.recordTransaction();
+	private TransactionImple addImportedTransaction(TransactionImple importedTransaction, Xid mapKey, Xid xid, int timeout) {
+		SubordinateXidImple importedXid = new SubordinateXidImple(mapKey);
+		AtomicReference<TransactionImple> holder = new AtomicReference<>(); // this class is an already-existing volatile field holder, might as well use it
+		AtomicReference<TransactionImple> existing;
 
-			return recovered;
+		if ((existing = _transactions.putIfAbsent(importedXid, holder)) != null) {
+			holder = existing;
 		}
-		else
-		{
-			return tx;
+
+		TransactionImple txn = holder.get();
+
+		if (txn == null) {
+			synchronized (holder) {
```

Comment:
I changed this a bit in https://github.com/mmusgrov/narayana/pull/2 - I think we can use the existing == null check here to reduce the number of holder.get calls